### PR TITLE
adding timeout attribute to function node

### PR DESCRIPTION
**Pull request by k1ln**
- [x] New feature (non-breaking change which adds functionality)

Discussion here: 
https://discourse.nodered.org/t/function-node-doesnt-have-timeout-feature/78483

## Proposed changes

Adding a timeout attribute to the function node, so an endless function doesnt break the node red server.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: packages/node_modules/@node-red/nodes/core/function/10-function.js
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/function/10-function.js:99]:
```suggestion
        node.timeout = n.timeout*1000;
```

File: packages/node_modules/@node-red/nodes/core/function/10-function.html
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/function/10-function.html:476]:
```suggestion
            // 4294967 is max in node.js timeout.
            $( "#node-input-timeout" ).spinner({
                min: 0,
                max: 4294967,
```

Change to second resolution.

File: packages/node_modules/@node-red/nodes/locales/de/messages.json
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/locales/de/messages.json:220]:
```suggestion
            "timeout": "Timeout"
```

File: packages/node_modules/@node-red/nodes/locales/en-US/messages.json
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/locales/en-US/messages.json:254]:
```suggestion
            "timeout": "Timeout"
```

File: packages/node_modules/@node-red/nodes/locales/ru/messages.json
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/locales/ru/messages.json:216]:
```suggestion
            "timeout":"Время ожидания"
```

File: test/nodes/core/function/10-function_spec.js
Comment (knolleary) [inline test/nodes/core/function/10-function_spec.js:1428]:
```suggestion
        var flow = [{id:"n1",type:"function",wires:[["n2"]],timeout:"0.010",func:"while(1==1){};\nreturn msg;"}];
```

File: packages/node_modules/@node-red/nodes/core/function/10-function.html
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/function/10-function.html:87]:
```suggestion
                <input id="node-input-timeout" style="width: 60px;" data-i18n="[placeholder]join.seconds">
```

Use a placeholder to indicate the units. This is consistent with other core nodes, such as the join node.

**Review by knolleary — CHANGES_REQUESTED** [review]
Thanks for the PR @k1ln - I've retargeted it to the `dev` branch as this is where new feature development is done. This has pulled some extra commits in and made a bit of muddle of this PR. My apologies - but we'll sort it out in the merge.

The basic behaviour is good. I think there are a couple changes to make that I have proposed changes for inline:

1. Use seconds rather than milliseconds for the value. Millisecond precision feels unnecessary for this. Having the spinner increment by 1ms intervals felt like a lot of clicking to do anything meaningful :)
2. Use the input placeholder to show the units rather than the label. This is consistent with other core nodes.

LMK if you're happy with these changes and we can get them merged.

This *might* miss 3.1.0-beta.3, but will certainly be ready for the final 3.1.0 release.

**Comment by k1ln** [comment]
> Thanks for the PR @k1ln - I've retargeted it to the `dev` branch as this is where new feature development is done. This has pulled some extra commits in and made a bit of muddle of this PR. My apologies - but we'll sort it out in the merge.
> 
> The basic behaviour is good. I think there are a couple changes to make that I have proposed changes for inline:
> 
> 1. Use seconds rather than milliseconds for the value. Millisecond precision feels unnecessary for this. Having the spinner increment by 1ms intervals felt like a lot of clicking to do anything meaningful :)
> 2. Use the input placeholder to show the units rather than the label. This is consistent with other core nodes.
> 
> LMK if you're happy with these changes and we can get them merged.
> 
> This _might_ miss 3.1.0-beta.3, but will certainly be ready for the final 3.1.0 release.

Can we have a setting which could be used like default time to wait for a funciton node? I don't exactly know how I can get or populate the settings from the settings.js in the app directory to the frontend where the node is. Should i make a new PR for the dev branch for this one? 

Kind Regards 

Kilian

**Comment by Steve-Mcl** [comment]
@k1ln from what I understand, this new feature, as it is currently implemented, shows an 0 value timeout by default (0 meaning no timeout)

I believe this is good enough for 1st version. It aligns with the current functionality pre V3.1 (as in no timeout by default). 

If you intend on implementing this, then I would suggest you branch off this branch and implement it as an extra feature. 

It may be that we have to wait until Node-RED v4 - as introducing a default value would be a (slightly) breaking change. That would of course need to be discussed on its merit and what that default value may look like.

**Comment by k1ln** [comment]
@Steve-Mcl ok. Do you perhaps know how one can propagate settings from the settings.js to the frontend. The implementation is not exactly sharp cause if I just add stuff in the seeting.js file I don't have them available in the Frontend on the fly what i would expect. Where do i need to change something to propagate settings from the settings.js file to the frontend.

**Comment by Steve-Mcl** [comment]
@k1ln 

Use [this doc](https://nodered.org/docs/creating-nodes/node-js#exposing-settings-to-the-editor) as a guide.


In essence: 

In `packages/node_modules/@node-red/nodes/core/function/10-function.js`, add 
```js
RED.nodes.registerType("function",FunctionNode, {
        dynamicModuleList: "libs",
        settings: {
            functionExternalModules: { value: true, exportable: true },
            functionTimeout: { value:30, exportable: true }, // << add
        }
    });
```


and in `packages/node_modules/node-red/settings.js`
image

then you can access `RED.settings.functionTimeout` in both the `.js` and `.html`

**Comment by k1ln** [comment]
@Steve-Mcl Thank you.

**Comment by dceejay** [comment]
As long as the default remains 0 (=infinite as today) - then allowing a setting to re-configure it could be useful.
though I would call it `functionNodeTimeout` to make it explicitly about that node.

**Comment by Steve-Mcl** [comment]
@k1ln could you rebase or merge dev into this branch and check there are no issues please?

**Comment by k1ln** [comment]
@Steve-Mcl didn't knolleary ödo that before? I pushed that button now but now it says it needs approval.

**Comment by k1ln** [comment]
@knolleary What do you mean? I resolved all your suggested changes above. Do i need to change anything more?

**Comment by knolleary** [comment]
@k1ln ah - you clicked 'resolve' without accepting the changes I proposed. No matter - I'll redo them.

**Comment by hardillb** [comment]
Just to be clear, clicking "Resolve" just closes the discussion, it does not change any code. You need to explicitly apply each of the suggested changes.

**Comment by knolleary** [comment]
I have unresolved the conversations and applied the changes. Once the tests pass we can get this merged. Thanks @k1ln

**Review by knolleary — APPROVED** [review]

**Comment by k1ln** [comment]
@knolleary Thank you next time i'll bei more careful.

**Pull request merged**
